ibis/backends/sql/rewrites.py

```diff
@@ -236,9 +236,14 @@ def sqlize(
     simplified = sqlized.replace(merge_select_select)
 
     # extract common table expressions while wrapping them in a CTE node
-    ctes = extract_ctes(simplified)
-    subs = {cte: CTE(cte) for cte in ctes}
-    result = simplified.replace(subs)
+    ctes = set(extract_ctes(simplified))
+
+    def wrap(node, _, **kwargs):
```

Comment:
Going to add this as comment, but the the problem with the previous solution was that `replace(dict)` matches on the original node and replaces the given value, but the ctes in the `subs` mapping are the original relation objects.

Instead we need to do the following: once we encounter a relation which is occurring more than once (`in ctes`) we need to wrap then in a `CTE` operation, which must propagate upwards in topological order, hence we are using a custom replacer which matches on the original (not the recreated) node.